File: crot.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>
#include <gsl/gsl_matrix.h>
#include <Imlib2.h>


#define VWIDTH 4.0
#define VHEIGHT 3.0
#define VX -2.5
#define VY -1.5

typedef unsigned int uint;

struct matrix
{
	int width, height;
	int **vals;
};


struct image
{
	int width, height;


	struct matrix *red;
	struct matrix *green;
	struct matrix *blue;
};

typedef struct image image;


struct matrix *newMatrix(int width, int height)
{
	int i;
	struct matrix *mat = (struct matrix *)calloc(1, sizeof(struct matrix));


	mat->vals = (int **)calloc(width, sizeof(int *));
	mat->width = width;
	mat->height = height;


	for(i = 0; i < width; i++)
	{
		mat->vals[i] = (int*)calloc(height, sizeof(int));
	}
	return mat;
}


image *loadImage(const char *filename)
{
	Imlib_Image img;
	int i, j, width, height;


	struct image *result = (struct image*)calloc(1, sizeof(struct image));


	img = imlib_load_image(filename);
	imlib_context_set_image(img);


	width = imlib_image_get_width();
	height = imlib_image_get_height();


	result->width = width;
	result->height = height;


	result->red = newMatrix(width, height);
	result->green = newMatrix(width, height);
	result->blue = newMatrix(width, height);


	for(i = 0; i < width; i++)
	{
		for(j = 0; j < height; j++)
		{
			Imlib_Color col;
			imlib_image_query_pixel(i, j, &col);


			result->red->vals[i][j] = (int)col.red;
			result->green->vals[i][j] = (int)col.green;
			result->blue->vals[i][j] = (int)col.blue;
		}
	}


	return result;
}


void saveImage(struct image *img, const char *filename)
{
	Imlib_Image out;
	int i, j;


	out = imlib_create_image(img->width, img->height);
	imlib_context_set_image(out);


	for(i = 0; i < img->width; i++)
	{
		for(j = 0; j < img->height; j++)
		{
			int red = (int)img->red->vals[i][j];
			int green = (int)img->green->vals[i][j];
			int blue = (int)img->blue->vals[i][j];


			red = red <= 255 ? red : 255;
			green = green <= 255 ? green : 255;
			blue = blue <= 255 ? blue : 255;


			imlib_context_set_color(red, green, blue, 255);
			imlib_image_draw_line(i, j, i, j, 0);
		}
	}


	imlib_save_image(filename);
}

uint mand(uint w, uint h, uint x, uint y){
	uint i;
	double complex z;
	double complex c;

	z = 0 + 0 * I;
	c = VX + VWIDTH/w * x + (VY+VHEIGHT/h * y) * I;

	for(i=0; i<256; i++){
		z = z*z+c;
		if(creal(z*conj(z)) > 20) break;
	}
	return i;
}

void rotx(image *img, float t,int x ,int y){
	int r,g,b;
	int r1,g1,b1;
	
	r1 = img->red->vals[x][y];
	g1 = img->green->vals[x][y];
	b1 = img->blue->vals[x][y];

	r = r1;
	g = cos(t)*g1 - sin(t)*b1;
	b = sin(t)*g1 + cos(t)*b1;
	
	img->red->vals[x][y] = abs(r);
	img->green->vals[x][y] = abs(g);
	img->blue->vals[x][y] = abs(b);
}


void rotz(image *img, float t,int x ,int y){
	int r,g,b;
	int r1,g1,b1;
	
	r1 = img->red->vals[x][y];
	g1 = img->green->vals[x][y];
	b1 = img->blue->vals[x][y];

	r = r1*cos(t) + sin(t) *g1;
	g = cos(t)*g1 - sin(t)*r1;
	b = b1;
	
	img->red->vals[x][y] = abs(r);
	img->green->vals[x][y] = abs(g);
	img->blue->vals[x][y] = abs(b);
}

void roty(image *img, float t,int x ,int y){
	int r,g,b;
	int r1,g1,b1;
	
	r1 = img->red->vals[x][y];
	g1 = img->green->vals[x][y];
	b1 = img->blue->vals[x][y];

	r = r1*cos(t) - sin(t) *b1;
	g = g1;
	b = b1*sin(t) + cos(t) * r1;
	
	img->red->vals[x][y] = abs(r);
	img->green->vals[x][y] = abs(g);
	img->blue->vals[x][y] = abs(b);
}

void rot(image *img, int x,int y,float r,float g,float b)
{
	rotx(img,r,x,y);
	roty(img,g,x,y);
	rotz(img,b,x,y);
}

int main(int argc, char**argv){

	int i,j;
	image *img;

	if(argc < 6){
		fprintf(stderr,"Usage: <input> <output> <mfr> <mfg> <mfb>\n");
		exit(0);
	}

	img = loadImage(argv[1]);

	float r,g,b;
	r = atof(argv[3]);
	g = atof(argv[4]);
	b = atof(argv[5]);
	
	for(i=0; i< img->width; i++){
		for(j=0; j<img->height; j++){
			uint m;
			m = mand(img->width,img->height,i,j);
			rot(img,i,j,m*r,m*g,m*b);
		}
	}
	saveImage(img,argv[2]);
	
}

File: crot.py
import Image
import sys
import math


def main():
	if len(sys.argv) < 6:
		print "Please name a file to rotate."
		sys.exit(0);

	img = Image.open(sys.argv[1]);
	pix = img.load();
	w,h = img.size;

	r = float(sys.argv[3]);
	g = float(sys.argv[4]);
	b = float(sys.argv[5]);


	for i in xrange(w):
		print i*1.0/w *100
		for j in xrange(h):
			
		#	r1 = dist(0,0,i,0)/1000.0 * r;
		#	g1 = dist(0,0,i,0)/1000.0 * g;
		#	b1 = dist(0,0,i,0)/1000.0 * b;
			m = mand(w,h,i,j);
			m=0 #/////////////////////////////////
			r1 = m/10+r;
			g1 = 0+g;
			b1 = 0+b;
			p = pix[i,j];
			pix[i,j] = rot(p,r1,g1,b1);
	img.save(sys.argv[2]);

def mand(w,h,x,y):
	a=0;
	b=0;
	c = -2.5+4.0/w*x
	d = -2+4.0/h*y
	for i in range(127):
		t = a
		a=a**2-b**2+c
		b=2*t*b+d
		if(a**2+b**2>10):
			break;
	return i;

def dist(a,b,c,d):
	return math.sqrt((a-c)**2+(b-d)**2);

def rotx(p,x):
	r1,g1,b1=p;
	r = r1;
	g = math.cos(x)*g1 - math.sin(x)*b1;
	b = math.sin(x)*g1 + math.cos(x) *b1;
	r = abs(r);
	g = abs(g);
	b = abs(b);
	p = r,g,b;
	return p;

def rotz(p,z):
	r1,g1,b1=p;
	r = r1*math.cos(z) + math.sin(z) *g1;
	g = -math.sin(z)*r1 + math.cos(z) *g1;
	b = b1;
	r = abs(r);
	g = abs(g);
	b = abs(b);
	p = r,g,b;
	return p;

def roty(p,y):
	r1,g1,b1=p;
	r = r1*math.cos(y) + math.sin(y) * b1;
	g = g1;
	b = -math.sin(y)*r1 + math.cos(y) *b1;
	r = abs(r);
	g = abs(g);
	b = abs(b);
	p = r,g,b;
	return p;

def rot(p,x,y,z):
	p = rotx(p,x);
	p = roty(p,y);
	p = rotz(p,z);
	r,g,b = p;
	r = int(r);
	g = int(g);
	b = int(b);
	p = r,g,b;
	return p;

main();	
